flytekit/core/workflow.py

```diff
@@ -277,87 +320,322 @@ def _local_execute(self, ctx: FlyteContext, **kwargs) -> Union[Tuple[Promise], P
             flyte_interface_types=self.interface.outputs,
             native_types=self.python_interface.outputs,
         )
-
+        # Recreate new promises that use the workflow's output names.
         new_promises = [Promise(var, wf_outputs_as_literal_dict[var]) for var in expected_output_names]
 
-        # TODO: With the native interface, create_task_output should be able to derive the typed interface, and it
-        #   should be able to do the conversion of the output of the execute() call directly.
-        return create_task_output(new_promises, self._native_interface)
+        return create_task_output(new_promises, self.python_interface)
+
+
+class ImperativeWorkflow(WorkflowBase):
+    def __init__(
+        self, name: str, failure_policy: Optional[WorkflowFailurePolicy] = None, interruptible: Optional[bool] = False,
+    ):
+        metadata = WorkflowMetadata(on_failure=failure_policy or WorkflowFailurePolicy.FAIL_IMMEDIATELY)
+        workflow_metadata_defaults = WorkflowMetadataDefaults(interruptible)
+        self._compilation_state = CompilationState(prefix="")
+        self._inputs = {}
+        self._unbound_inputs = set()
+        super().__init__(
+            name=name,
+            workflow_metadata=metadata,
+            workflow_metadata_defaults=workflow_metadata_defaults,
+            python_interface=Interface(),
+        )
+
+    @property
+    def compilation_state(self) -> CompilationState:
+        """
+        Compilation is done a bit at a time, one task or other entity call at a time. This is why this workflow
+        class has to keep track of its own compilation state.
+        """
+        return self._compilation_state
+
+    @property
+    def nodes(self) -> List[Node]:
+        return self._compilation_state.nodes
+
+    @property
+    def inputs(self) -> Dict[str, Promise]:
+        """
+        This holds the input promises to the workflow. The nodes in these Promise objects should always be from
```

Comment:
nit: The **nodes** in these Promise objects should always be from (?)

Reply:
huh?

Reply:
I'll reword a bit, but not sure why "nodes" was in bold.

Reply:
oh sorry i was confused by this comment